indexer-balances/src/db_adapters/balance_changes.rs

```diff
@@ -581,121 +560,35 @@ fn get_deltas(
+    block_height: u64,
+    balance_cache: &cache::BalanceCache,
+    pool: &sqlx::Pool<sqlx::Postgres>,
 ) -> anyhow::Result<crate::BalanceDetails> {
-    let mut balances_cache_lock = balance_cache.lock().await;
-    let result = match balances_cache_lock.cache_get(account_id) {
-        None => {
-            let account_balance =
-                match get_account_view(json_rpc_client, account_id, block_hash).await {
-                    Ok(account_view) => Ok(crate::BalanceDetails {
-                        non_staked: account_view.amount,
-                        staked: account_view.locked,
-                    }),
-                    Err(err) => match err.handler_error() {
-                        Some(RpcQueryError::UnknownAccount { .. }) => Ok(crate::BalanceDetails {
-                            non_staked: 0,
-                            staked: 0,
-                        }),
-                        _ => Err(err.into()),
-                    },
-                };
-            if let Ok(balance) = account_balance {
-                balances_cache_lock.cache_set(account_id.clone(), balance);
-            }
-            account_balance
-        }
-        Some(balance) => Ok(*balance),
-    };
-    drop(balances_cache_lock);
-    result
-}
+    if let Some(balance) = balance_cache.get(account_id).await {
+        return Ok(balance);
+    }
 
-async fn save_latest_balance(
-    account_id: near_indexer_primitives::types::AccountId,
-    balance: &crate::BalanceDetails,
-    balance_cache: &crate::BalanceCache,
-) {
-    let mut balances_cache_lock = balance_cache.lock().await;
-    balances_cache_lock.cache_set(
-        account_id,
-        crate::BalanceDetails {
-            non_staked: balance.non_staked,
-            staked: balance.staked,
+    let account_balance = match crate::models::select_one_retry_or_panic(
+        pool,
+        &NearBalanceEvent::select_prev_balance_query(block_height, account_id),
+        crate::RETRY_COUNT,
+    )
+    .await
+    {
+        Ok(Some(balance_event)) => crate::BalanceDetails {
+            non_staked: u128::from_str(&balance_event.absolute_nonstaked_amount.to_string())?,
+            staked: u128::from_str(&balance_event.absolute_staked_amount.to_string())?,
         },
-    );
-    drop(balances_cache_lock);
-}
-
-async fn get_account_view(
-    json_rpc_client: &near_jsonrpc_client::JsonRpcClient,
-    account_id: &near_indexer_primitives::types::AccountId,
-    block_hash: &near_indexer_primitives::CryptoHash,
-) -> Result<near_indexer_primitives::views::AccountView, JsonRpcError<RpcQueryError>> {
-    let query = near_jsonrpc_client::methods::query::RpcQueryRequest {
-        block_reference: near_primitives::types::BlockReference::BlockId(
-            near_primitives::types::BlockId::Hash(*block_hash),
-        ),
-        request: near_primitives::views::QueryRequest::ViewAccount {
-            account_id: account_id.clone(),
+        Ok(None) => crate::BalanceDetails {
```

Comment:
I'm not sure if returning `0` is the correct thing to do here but _think_ it's ok.

If we trust that the we have indexed all previous data, which we have, we should only encounter this branch _if_ the account doesn't exist, meaning this is the right thing to return.

Please correct me if I'm wrong!

Reply:
Your thoughts are correct only if we collect the data sequentially.
One more issue of this proposal - with these changes, we can't run several instances of the indexer to backfill the data faster. 
We always relied on this.

Reply:
Could you please think about all the options we have here?
I'd be happy to have a call and brainstorm with you together